File: 85-maximal-rectangle/maximal-rectangle.java
class Solution {
    public int maximalRectangle(char[][] matrix) {
        
        int max_area = 0;
        int rows = matrix.length;
        int cols = matrix[0].length;
        int n = cols;
        int[] heights = new int[cols];

        for(int i=0 ; i<rows ; i++){
            for(int j=0 ; j<cols ; j++){

                if(matrix[i][j] == '0'){
                    heights[j] = 0;
                }
                else{
                    heights[j]++;
                }

            }

            Stack<Integer> stack = new Stack<>();

            for(int j=0 ; j<cols ; j++){

                while(!stack.empty() && heights[j]<heights[stack.peek()]){
                    int curr_height = heights[stack.peek()];
                    stack.pop();
                    int NSR = j;
                    int NSL;

                    if(stack.empty()){
                        NSL = -1;
                    }
                    else{
                        NSL = stack.peek();
                    }

                    int curr_area = curr_height * ( NSR - NSL - 1 );
                    if(curr_area > max_area){
                        max_area = curr_area;
                    }

                }
                stack.push(j);

            }

            while(!stack.empty()){
                int curr_height = heights[stack.peek()];
                stack.pop();
                int NSR = n;
                int NSL;
                if(stack.empty()){
                    NSL = -1;
                }
                else{
                    NSL = stack.peek();
                }

                int curr_area = curr_height * (NSR - NSL - 1);
                if(curr_area > max_area){
                    max_area = curr_area;
                }

            }

        }

        return max_area;

    }
}

File: 85-maximal-rectangle/maximal-rectangle.py
class Solution:
    def maximalRectangle(self, matrix: list[list[str]]) -> int:
        
        rows = len(matrix)
        cols = len(matrix[0])
        n = cols
        max_area = 0

        heights = [0] * cols

        for row in range(rows):
            for col in range(cols):
                if(matrix[row][col] == '0'):
                    heights[col] = 0
                else:
                    heights[col]+=1
            
            stack = []

            for i in range(n):
                while(stack and heights[i]<heights[stack[-1]]):

                    curr_height = heights[stack[-1]]
                    stack.pop()

                    NSR = i
                    NSL = -1
                    if stack:
                        NSL = stack[-1]
                    
                    max_area = max(
                        max_area,
                        curr_height * (NSR - NSL - 1)
                    )
                stack.append(i)
            while(stack):
                
                curr_height = heights[stack[-1]]
                stack.pop()
                NSR = n
                NSL = -1
                if stack:
                    NSL = stack[-1]
                max_area = max(
                    max_area,
                    curr_height * (NSR - NSL - 1)
                )

        return max_area
